client/src/use/useAttributes.ts

```diff
@@ -0,0 +1,93 @@
+
+import {
+  ref, Ref, computed, set as VueSet, del as VueDel,
+} from '@vue/composition-api';
+
+export interface Attribute {
+  belongs: 'track' | 'detection';
+  datatype: 'text' | 'number' | 'boolean';
+  values?: string[];
+  name: string;
+  _id: string;
+}
+
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+export function isAttribute(obj: any): obj is Attribute {
+  return (
+    (typeof obj === 'object')
+    && (typeof obj.belongs === 'string')
+    && (typeof obj.datatype === 'string')
+    && (typeof obj._id === 'string')
+    && (typeof obj.name === 'string')
+  );
+}
+
+export type Attributes = Record<string, Attribute>;
+
+/**
+ * Modified markChangesPending for attributes specifically
+ */
+interface UseAttributesParams {
+  markChangesPending: (
+    {
+      action,
+      data,
+    }: {
+      action: 'upsert' | 'delete';
+      data: Attribute;
+    }
+  ) => void;
+}
+
+export default function UseAttributes({ markChangesPending }: UseAttributesParams) {
+  const attributes: Ref<Record<string, Attribute>> = ref({});
+
+
+  function loadAttributes(metadataAttributes: Record<string, Attribute>) {
+    attributes.value = metadataAttributes;
+  }
+
+  const getAttributes = computed(() => Object.values(attributes.value));
+
+  function setAttribute({ data }: {data: Attribute }, updateAllTracks = false) {
+    let oldAttribute;
+    if (data._id === '') {
+      // eslint-disable-next-line no-param-reassign
+      data._id = `${data.belongs}_${data.name}`;
```

Comment:
Sometimes `no-param-reassign` is a nuisance, but here I think it's protecting you.  This function has an undocumented side-effect in that it sets the `_id` of the passed attribute.  The caller relies on this behavior in an unclear way.

I actually see `_id` as unnecessary.  It's synthesized from other properties and doesn't really provide any information.  You could get rid of `_id` completely, have the map key be an implementation detail of `useAttributes`, and change the signature of `deleteAttirbute` to also accept the full object rather than an id.

Reply:
I'm going to change the `setAttribute` and `deleteAttribute` to accept the full object.  I started removing the `_id` from all Attributes and it just felt really wrong after a while computing the index each time I wanted to access it.  So I think I'll rename the `_id` to `key` and prevent any reassignment.

I think it might be better to have `setAttribute` be able to accept an oldAttribute if one exists (for the instance of renaming the attribute and changing the index).  So if a rename occurs it will delete the oldAttribute and replace it with the new one with the updated index.  There are already client side checks to make sure you don't create a new attribute with the same `key` as existing ones.

Reply:
I'm fine with `key`, but that leaves the side-effect issue.  An option is to have `setAttribute` return a clone with `key` set such that using it requires `localReactiveCopy.value = setAttribute(localReactiveCopy.value);`?

Reply:
The `data` Attribute isn't a direct reference, it's a new Attribute item built up of the changes and the `oldAttribute` is a reference to the item in the list and is only included if the user is updating an existing attribute.
The `oldAttribute` can be updated in the process by the name change causing a delete and refresh or by changing the type.  When this occurs the item that had a reference to that as being open is already closed (`AttributeEditor.vue`) and as  as the `VueSet` updates the attributes list it is then updated in the `trackDetails.vue`.  The value isn't returned and instead relies on the reactivity of the `attributes` list provided to `trackDetails.vue` (Not ideal, but who is going to have an extremely large list on one dataset).

![image](https://user-images.githubusercontent.com/61746913/109706576-1fc84a80-7b67-11eb-9714-cbced762f0dc.png)

Please let me know if I'm missing something here and seeing stuff incorrectly.

Reply:
I'm talking about the invocation from `TrackDetailsPanel.vue`, line 130.

```ts
      try {
        // before: saveData._id === ''
       await setAttribute({ data: saveData.data });
        // after: saveData._id === 'Belongs_NewName';
      } catch (err) {
        editingError.value = err.message;
      }
```

That `_id` change on `saveData` is a side-effect I'd like to avoid.  It looks like you may have already changed this, so I'll review once you push.